### PR TITLE
Enable via for lily58

### PR DESCRIPTION
**Pull request by gwillad**
## Description

adds keymap for lily58 that can be used in VIA 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/lily58/keymaps/via/keymap.c [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+#ifdef PROTOCOL_LUFA
+  #include "lufa.h"
+  #include "split_util.h"
+#endif
+#ifdef SSD1306OLED
+  #include "ssd1306.h"
+#endif
+
+extern uint8_t is_master;
+
```
Comment (drashna) [inline keyboards/lily58/keymaps/via/keymap.c:12]: This shouldn't be needed. 

```suggestion
```

File: keyboards/lily58/keymaps/via/keymap.c [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+
+ [_QWERTY] = LAYOUT( \
+  KC_ESC,   KC_1,   KC_2,    KC_3,    KC_4,    KC_5,                     KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_GRV, \
+  KC_TAB,   KC_Q,   KC_W,    KC_E,    KC_R,    KC_T,                     KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_MINS, \
+  KC_LCTRL, KC_A,   KC_S,    KC_D,    KC_F,    KC_G,                     KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, \
+  KC_LSFT,  KC_Z,   KC_X,    KC_C,    KC_V,    KC_B, KC_LBRC,  KC_RBRC,  KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,  KC_RSFT, \
+                        KC_LALT, KC_LGUI, MO(_LOWER), KC_SPC, KC_ENT, MO(_RAISE), KC_BSPC, KC_RGUI \
```
Comment (drashna) [inline keyboards/lily58/keymaps/via/keymap.c:42]: Via supports tri layer, if you use the macros for it. 

```suggestion
                        KC_LALT, KC_LGUI, FN_MO12, KC_SPC, KC_ENT, FN_MO23, KC_BSPC, KC_RGUI \
```

File: keyboards/lily58/keymaps/via/keymap.c [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+};
+
+// Setting ADJUST layer RGB back to default
+void update_tri_layer_RGB(uint8_t layer1, uint8_t layer2, uint8_t layer3) {
+  if (IS_LAYER_ON(layer1) && IS_LAYER_ON(layer2)) {
+    layer_on(layer3);
+  } else {
+    layer_off(layer3);
+  }
+}
```
Comment (drashna) [inline keyboards/lily58/keymaps/via/keymap.c:117]: This isn't used, nor is needed. 

```suggestion
```

File: keyboards/lily58/keymaps/via/keymap.c [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    // If you want to change the display of OLED, you need to change here
+    oled_write_ln(read_layer_state(), false);
+    oled_write_ln(read_keylog(), false);
+    oled_write_ln(read_keylogs(), false);
+    //oled_write_ln(read_mode_icon(keymap_config.swap_lalt_lgui), false);
+    //oled_write_ln(read_host_led_state(), false);
+    //oled_write_ln(read_timelog(), false);
+  } else {
+    oled_write(read_logo(), false);
+  }
```
Comment (drashna) [inline keyboards/lily58/keymaps/via/keymap.c:151]: Honestly, I would much rather that this be converted to use `oled_write_ln_P`, which means that you'd have to remve the SRC for the /lib/ files in the rules.mk, and move the functions into this file. 

This is more work, but is important, IMO, because the `*_P` functions use a lot less memory, which is a factor on split boards with OLED (source, me, as I've run into this issue personally, without adding VIA into the mix).

File: keyboards/lily58/keymaps/via/rules.mk [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
+
+# If you want to change the display of OLED, you need to change here
+SRC +=  ./lib/rgb_state_reader.c \
+        ./lib/layer_state_reader.c \
+        ./lib/logo_reader.c \
+        ./lib/keylogger.c \
+        # ./lib/mode_icon_reader.c \
+        # ./lib/host_led_state_reader.c \
+        # ./lib/timelogger.c \
```
Comment (drashna) [inline keyboards/lily58/keymaps/via/rules.mk:32]: This should only need: 
```suggestion
OLED_DRIVER_ENABLE= yes     # OLED display
VIA_ENABLE = yes

```

File: keyboards/lily58/keymaps/via/config.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+#define MASTER_LEFT
+// #define MASTER_RIGHT
+// #define EE_HANDS
+
+// #define SSD1306OLED
+
+#define USE_SERIAL_PD2
+
+#define TAPPING_FORCE_HOLD
```
Comment (drashna) [inline keyboards/lily58/keymaps/via/config.h:35]: I'm not a fan of having this defined by default, as it can break some keycodes.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/lily58/keymaps/via/keymap.c (hunk 4 above)
Comment (gwillad) [reply keyboards/lily58/keymaps/via/keymap.c:135]: i have not yet addressed this, although I took a stab at it. In my efforts to do so, i was able to reduce firmware size from about 80% to 76%, but i was unable to get the keylogger to work correctly. Layer and logo went over fine.

Comment (drashna) [reply keyboards/lily58/keymaps/via/keymap.c]: The keylogger won't work with the `*_P` stuff.  The `*_P` is to denote that the strings are in PROGMEM (program memory, aka firmware) rather than in memory.  However, the keylogger MUST be in memory, since the data changes, as PROGMEM cannot be changed (at least, not without flashing). 

So I don't think the "const" is strictly needed, but since the data can't/won't change, it's better to use const. 

Basically, anythign that needs to dynamically change (such as the keylogger stuff) can't be progmem, but everything else can and should be.

Comment (drashna) [reply keyboards/lily58/keymaps/via/keymap.c]: Also, if you want a good example of what you can do with the OLED stuff, I have a good example in the corne keymap folder: 
https://github.com/qmk/qmk_firmware/blob/master/keyboards/crkbd/keymaps/oled_sample/keymap.c

**Review by drashna — APPROVED** [review]

File: keyboards/lily58/keymaps/via/keymap.c (hunk 4 above)
Comment (gwillad) [reply keyboards/lily58/keymaps/via/keymap.c]: in that case would you prefer that I leave the keylogger as not *_P? or simply remove it?

File: keyboards/lily58/keymaps/via/keymap.c [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+
+ [_QWERTY] = LAYOUT( \
+  KC_ESC,   KC_1,   KC_2,    KC_3,    KC_4,    KC_5,                     KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_GRV, \
+  KC_TAB,   KC_Q,   KC_W,    KC_E,    KC_R,    KC_T,                     KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_MINS, \
+  KC_LCTRL, KC_A,   KC_S,    KC_D,    KC_F,    KC_G,                     KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, \
+  KC_LSFT,  KC_Z,   KC_X,    KC_C,    KC_V,    KC_B, KC_LBRC,  KC_RBRC,  KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,  KC_RSFT, \
+                        KC_LALT, KC_LGUI, FN_MO13, KC_SPC, KC_ENT, FN_MO23, KC_BSPC, KC_RGUI \
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/via/keymap.c:33]:
```suggestion
 [_QWERTY] = LAYOUT(
  KC_ESC,   KC_1,   KC_2,    KC_3,    KC_4,    KC_5,                     KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_GRV,
  KC_TAB,   KC_Q,   KC_W,    KC_E,    KC_R,    KC_T,                     KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_MINS,
  KC_LCTRL, KC_A,   KC_S,    KC_D,    KC_F,    KC_G,                     KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,
  KC_LSFT,  KC_Z,   KC_X,    KC_C,    KC_V,    KC_B, KC_LBRC,  KC_RBRC,  KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,  KC_RSFT,
                        KC_LALT, KC_LGUI, FN_MO13, KC_SPC, KC_ENT, FN_MO23, KC_BSPC, KC_RGUI
```

File: keyboards/lily58/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+ *                   | LAlt | LGUI |LOWER | /Space  /       \Enter \  |RAISE |BackSP| RGUI |
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+[_LOWER] = LAYOUT( \
+  _______, _______, _______, _______, _______, _______,                   _______, _______, _______,_______, _______, _______,\
+  KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,                     KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12, \
+  KC_GRV, KC_EXLM, KC_AT,   KC_HASH, KC_DLR,  KC_PERC,                   KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, KC_TILD, \
+  _______, _______, _______, _______, _______, _______, _______, _______, XXXXXXX, KC_UNDS, KC_PLUS, KC_LCBR, KC_RCBR, KC_PIPE, \
+                             _______, _______, _______, _______, _______,  _______, _______, _______\
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/via/keymap.c:54]:
```suggestion
[_LOWER] = LAYOUT(
  _______, _______, _______, _______, _______, _______,                   _______, _______, _______,_______, _______, _______,
  KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,                     KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,
  KC_GRV, KC_EXLM, KC_AT,   KC_HASH, KC_DLR,  KC_PERC,                   KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, KC_TILD,
  _______, _______, _______, _______, _______, _______, _______, _______, XXXXXXX, KC_UNDS, KC_PLUS, KC_LCBR, KC_RCBR, KC_PIPE,
                             _______, _______, _______, _______, _______,  _______, _______, _______
```

File: keyboards/lily58/keymaps/via/keymap.c [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+
+[_RAISE] = LAYOUT( \
+  _______, _______, _______, _______, _______, _______,                     _______, _______, _______, _______, _______, _______, \
+  KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,                        KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    _______, \
+  KC_F1,  KC_F2,    KC_F3,   KC_F4,   KC_F5,   KC_F6,                       XXXXXXX, KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT, XXXXXXX, \
+  KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,   _______, _______,  KC_PLUS, KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC, KC_BSLS, \
+                             _______, _______, _______,  _______, _______,  _______, _______, _______ \
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/via/keymap.c:76]:
```suggestion
[_RAISE] = LAYOUT(
  _______, _______, _______, _______, _______, _______,                     _______, _______, _______, _______, _______, _______,
  KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,                        KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    _______,
  KC_F1,  KC_F2,    KC_F3,   KC_F4,   KC_F5,   KC_F6,                       XXXXXXX, KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT, XXXXXXX,
  KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,   _______, _______,  KC_PLUS, KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC, KC_BSLS,
                             _______, _______, _______,  _______, _______,  _______, _______, _______
```

File: keyboards/lily58/keymaps/via/keymap.c [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+ *                   | LAlt | LGUI |LOWER | /Space  /       \Enter \  |RAISE |BackSP| RGUI |
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+  [_ADJUST] = LAYOUT( \
+  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, \
+  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, \
+  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, \
+  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,\
+                             _______, _______, _______, _______, _______,  _______, _______, _______ \
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/via/keymap.c:97]:
```suggestion
  [_ADJUST] = LAYOUT(
  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
                             _______, _______, _______, _______, _______,  _______, _______, _______
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/lily58/keymaps/via/keymap.c (hunk 4 above)
Comment (drashna) [reply keyboards/lily58/keymaps/via/keymap.c]: The keylogger must not use the `_P` version.  That is fine, here.

Comment (gwillad) [reply keyboards/lily58/keymaps/via/keymap.c]: okay, leaving the keylogger using _P functions. Currently compiles to use about 80% of available space

**Comment by gwillad** [comment]
[someone](https://github.com/qmk/qmk_firmware/pull/10178) beat me to it :)

**Comment by elfmimi** [comment]
someone? it's the designer of lily58 oneself that who beat you to it. @gwillad

**Comment by gwillad** [comment]
> someone? it's the designer of lily58 oneself that who beat you to it. @gwillad

hahaha shows what i know!
